GEXF Renderer

lindenb opened the pull request:
Hi All,

as discussed on slack ( https://app.slack.com/client/TE6CZUZPH/C04AQ3LHJUF ) here is my PR to add a GEXF renderer to nf-prov.

I'm mostly a java dev, may be there are some more "groovy ways" to do this.

P.

lindenb committed:
gexf renderer

lindenb committed:
gexf renderer

bentsherman commented:
Thanks Pierre. The implementation looks good overall. I will clean it up a bit and merge when I have some time, hopefully by the end of the year.

I was thinking to add this as a sub-format of the `dag` format. But then you wouldn't be able to output both GEXF and Mermaid in the same run, which is nice to have for the e2e test.

You could argue that all of these prov formats are just different ways of showing the DAG. So maybe I'll just rename the `dag` format to `mermaid`, in a separate PR.

bentsherman committed:
cleanup

bentsherman commented on src/main/groovy/nextflow/prov/renderers/GexfRenderer.groovy:
```diff
@@ -0,0 +1,531 @@
+class GexfRenderer implements Renderer {
+
+    private static final String VERSION = "1.3"
+    private static final String XMLNS = "http://www.gexf.net/1.3"
```
I'm getting an error in VS Code (using the Red Hat XML extension) for the xmlns:

```
TargetNamespace.1: Expecting namespace 'http://www.gexf.net/1.3', but the target namespace of the schema document is 'http://gexf.net/1.3'.
```

Doesn't seem to be an issue for Gephi, but VS Code can't recognize the `gexf` element. Not a big deal, but if you know of an easy fix then we might as well do it

bentsherman commented on src/main/groovy/nextflow/prov/renderers/GexfRenderer.groovy:
```diff
@@ -0,0 +1,531 @@
+        long getFileTimestamp() {
+            if( !path.isFile() ) return -1L
+            try {
+                final fileTime = Files.readAttributes(path, BasicFileAttributes.class).creationTime()
```
Do you think it would be better to use the creation time or last modified time? I don't have a hard opinion on this but it seems the last modified time would be more consistent

bentsherman commented on src/main/groovy/nextflow/prov/renderers/GexfRenderer.groovy:
```diff
@@ -0,0 +1,531 @@
+            w.writeStartElement("meta")
+            w.writeAttribute("lastmodifieddate", new SimpleDateFormat("yyyy-MM-dd").format(new Date()))
+            w.writeStartElement("creator")
+            w.writeCharacters("GEXF Renderer for nf-prov by Pierre Lindenbaum")
```
Are you sure you want your name in every GEXF file created by every nextflow run around the world? :sweat_smile:

bentsherman reviewed (commented):
Fortunately we already try not to use too much Groovy magic in the runtime, so your style is pretty close :smile: 

I did some cleanup and left some comments for you. As far as I can tell the e2e test looks good, but let me know if everything still works on your end

bentsherman committed:
minor edits

lindenb committed:
changed xml namespace, removed my name, added modified+created timestamp

lindenb commented:
@bentsherman sorry for the delay, I was away from my code. Thanks for looking at this code. Back to you. I removed my name, added both created/modified attributes. I also fixed the xmlns and it looks like it's still working in gephi online: 
<img width="1847" height="828" alt="jeter" src="https://github.com/user-attachments/assets/f8e4ab6d-fdb8-42af-8fca-cc4a6880db9f" />

lindenb committed:
fix filetype=file

bentsherman committed:
minor edits
